src/mlpack/core/optimizers/saga/saga_impl.hpp

```diff
@@ -90,67 +99,72 @@ double SAGAType<UpdatePolicyType, DecayPolicyType>::Optimize(
   {
-    // Calculate the objective function.
-    overallObjective = 0;
-    for (size_t f = 0; f < numFunctions; f += batchSize)
-    {
-      const size_t effectiveBatchSize = std::min(batchSize, numFunctions - f);
-      overallObjective += function.Evaluate(iterate, f, effectiveBatchSize);
-    }
-
-    if (std::isnan(overallObjective) || std::isinf(overallObjective))
+    // Is this iteration the start of a sequence?
+    if ((currentFunction % numFunctions) == 0)
     {
-      Log::Warn << "SAGA: converged to " << overallObjective
-                << "; terminating  with failure.  Try a smaller step size?"
-                << std::endl;
-      return overallObjective;
-    }
 
-    if (std::abs(lastObjective - overallObjective) < tolerance)
-    {
-      Log::Info << "SAGA: minimized within tolerance " << tolerance
-                << "; terminating optimization." << std::endl;
-      return overallObjective;
+      if (std::isnan(overallObjective) || std::isinf(overallObjective))
+      {
+        Log::Warn << "SAGA: converged to " << overallObjective
+                  << "; terminating  with failure.  Try a smaller step size?"
+                  << std::endl;
+        return overallObjective;
+      }
+
+      if (std::abs(lastObjective - overallObjective) < tolerance)
+      {
+        Log::Info << "SAGA: minimized within tolerance " << tolerance
+                  << "; terminating optimization." << std::endl;
+        return overallObjective;
+      }
+
+      // Reset the counter variables.
+      lastObjective = overallObjective;
+      overallObjective = 0;
+      currentFunction = 0;
+
+      if (shuffle) // Determine order of visitation.
+        function.Shuffle();
     }
 
-    lastObjective = overallObjective;
-
     // Store current parameter for the calculation of the variance reduced
     // gradient.
     iterate0 = iterate;
 
-    for (size_t f = 0, b=0, currentFunction = 0; f < numFunctions;
-      /* incrementing done manually */)
-    {
-      b = math::RandInt(0, numBatches); // Select a random batch
-
-      // Find the effective batch size (the last batch may be smaller).
-      currentFunction = b * batchSize;
-      effectiveBatchSize = std::min(batchSize, numFunctions - currentFunction);
+    batch = math::RandInt(0, numBatches); // Select a random batch
 
-      // Calculate the gradient of a random function.
-      function.Gradient(iterate, currentFunction, gradient,
-                        effectiveBatchSize);
-      gradient0 = tableOfGradients.slice(b);
+    // Find the effective batch size; we have to take the minimum of three
+    // things:
+    // - the batch size can't be larger than the user-specified batch size;
+    // - the batch size can't be larger than the number of iterations left
+    //       before actualMaxIterations is hit;
+    // - the batch size can't be larger than the number of functions left.
+    effectiveBatchSize = std::min(
```

Comment:
This should use ``batch``instead of ``i`` right?

Reply:
I don't think so! In implementations like `sgd` we use `i`